lib/src/kbasesearchengine/main/IndexerCoordinator.java

```diff
@@ -17,33 +20,100 @@
 import kbasesearchengine.events.storage.StatusEventStorage;
 import kbasesearchengine.tools.Utils;
 
+/** Coordinates which events will get processed by {@link IndexerWorker}s based on the
+ * {@link EventQueue}. The responsibility of the coordinator is to periodically update the event
+ * state in the {@link StatusEventStorage} such that the workers process the correct events.
+ * 
+ * Only one indexer coordinator should run at one time.
+ * 
+ * This class is not thread safe.
+ * @author gaprice@lbl.gov
+ *
+ */
 public class IndexerCoordinator {
     
     private static final int RETRY_COUNT = 5;
     private static final int RETRY_SLEEP_MS = 1000;
-    private static final List<Integer> RETRY_FATAL_BACKOFF_MS = Arrays.asList(
+    private static final List<Integer> RETRY_FATAL_BACKOFF_MS_DEFAULT = Arrays.asList(
             1000, 2000, 4000, 8000, 16000);
     
     private final StatusEventStorage storage;
     private final LineLogger logger;
-    private ScheduledExecutorService executor = null;
+    private final ScheduledExecutorService executor;
+    private final EventQueue queue;
     
-    private final Retrier retrier = new Retrier(RETRY_COUNT, RETRY_SLEEP_MS,
-            RETRY_FATAL_BACKOFF_MS,
-            (retrycount, event, except) -> logError(retrycount, event, except));
+    private final int maxQueueSize;
+    private int continuousCycles = 0;
+    
+    private final Retrier retrier;
 
+    /** Create the indexer coordinator. Only one coordinator should run at one time.
+     * @param storage the storage system containing events.
+     * @param logger a logger.
+     * @param maximumQueueSize the maximum number of events in the internal in-memory queue.
+     * This should be a fairly large number because events may not arrive in the storage system
+     * in the ordering of their timestamps, and so the queue acts as a buffer so events can be
+     * sorted before processing if an event arrives late.
+     * @throws InterruptedException if the thread is interrupted while attempting to initialize
+     * the coordinator.
+     * @throws IndexingException if an exception occurs while trying to initialize the
+     * coordinator.
+     */
+    public IndexerCoordinator(
+            final StatusEventStorage storage,
+            final LineLogger logger,
+            final int maximumQueueSize)
+            throws InterruptedException, IndexingException {
+        this(storage, logger, maximumQueueSize, Executors.newSingleThreadScheduledExecutor(),
+                RETRY_FATAL_BACKOFF_MS_DEFAULT);
+    }
+    
+    /** Create an indexer coordinator solely for the purposes of testing. This constructor should
+     * not be used for any other purpose.
+     * @param storage the storage system containing events.
+     * @param logger a logger.
+     * @param maximumQueueSize the maximum number of events in the internal in-memory queue.
+     * @param testExecutor a single thread executor for testing purposes, usually a mock.
```

Comment:
We have to inject the mocks somehow, and it seems to me a test constructor is as good as anything else. What I've done in the past is make the test constructor private and create test instances with reflection, but my feeling at this point is that that's overly complicated and just having a public constructor is ok. If this were library code I probably would make it private, but since most likely no one except for a few KBase devs will ever touch it I'm not worried about it.

Anyway, happy to make the constructor private if you feel strongly about it or if you have some other suggestion, suggest away.